eip-5564.md

```diff
@@ -8,174 +8,170 @@ status: Draft
-    ///  Senders can use this public key for stealth address generation.
-    /// @param registrant The address of the registrant
-    /// @param pubKeyPrefix The public key prefix
-    /// @param pubKey The public key
-    function registerKey(address _registrant, uint256 pubKeyPrefix, uint256 pubKey) 
-    
-    
-    /// @notice Returns the stealth key associated with an address.
-    /// @param registrant The address whose keys to lookup.
-    /// @return pubKeyPrefix X The public key prefix
-    /// @return pubKey The public key
-    function stealthKeys(address registrant) returns (uint256 PublicKeyX, uint256 PublicKeyY)
-    
-
-    /// @notice Generates Stealth Address on behalve of a registered user.
-    /// @dev The caller executes this function locally to compute a stealth address
-    ///   that can be accessed by the recipient.
-    /// Further, function computes the announcement S that can be published.
-    /// @param secret A secret value selected by the sender
-    /// @param recipient The address of the recipient
-    function generateStealthAddress(uint256 secret, address recipient) returns (bytes announcement, address stealthAddress)
-
-    /// @notice Emits event with information on registered keys.
-    /// @dev Event emitted when a user sets/updates their registered public key.
-    /// @param registrant The address of the registrant
-    /// @param pubKeyPrefix The public key prefix
-    /// @param pubKey The public key
-    event StealthKeyChanged(address indexed registrant, uint256 pubKeyPrefix, uint256 pubKey);
-    
-    
-    /// @notice Emits event with information on the stealth-address transfer.
-    /// @dev Event contains the announcement S and the recipient's address.
-    ///  S is generated by the sender and represents the public key to the secret s.
-    ///  The sender broadcasts S for every stealth-address transfer. Users can use S to check 
-    ///  if they were the recipients of a respective transfer by comparing it to stealthRecipient.
-    /// @param stealthRecipient The address to send the funds to
-    /// @param announcement The public key to the sender's secret
-    event Annoucement(address indexed stealthRecipient, bytes announcement)
+pragma solidity ^0.8.0;
+
+/// @notice Registry mapping an address to its stealth key information.
+interface IERC5564Registry {
+  /// @notice Returns the stealth public keys for the given `registrant` to compute a stealth
+  /// address accessible only to that `registrant` using the provided `generator` contract.
+  /// @dev MUST return zero if a registrant has not registered keys for the given generator.
+  /// @dev The pubkeys are returned as `bytes memory` to support pubkeys of any  format and size.
+  /// @dev Generator contracts SHOULD be written to support a single public key, or two public keys:
+  /// one viewing key and one spending key. This can be inferred from the length of the key material.
```

Comment:
One discussion we had was: should the spending vs. viewing key distinction be part of the EIP? I learn towards yes, here's why—we can either:
1. Include the distinction in this EIP, and users can set the spending vs. viewing keys equal to each other if they don't care about the distinction (that's the stance this PR takes).
2. Or we can not include the distinction in this EIP, and have a second EIP that extends this and adds support for it.

The first option seems preferable to me: Given that scanning time is still an unsolved problem, as stealth addresses gain popularity it's likely people may want to outsource scanning to a service. But with the second option, it's possible that integrators or contracts implement stealth addresses _without_ the spend vs. view key distinction, and then it fragments things if we have e.g. two registries, or two generator contracts. Additionally, the cost to having this distinction in the EIP is fairly small IMO.

One question I'm not sure of is: Will all possible schemes, e.g. quantum resistant ones, also support this spend vs. view key distinction?

Reply:
I agree and would also lean towards 1. And as you said, it can still be used with a single key if preferred.
Still, it would be beneficial to have a scanning app that users who prefer the one key solution can run locally while connected to a custom rpc. Then users could download all Announcements from some central place and retrieve their stealth private keys without having to input their key on some online site.

On the other hand, 2 keys adds some complexity and could possible result in users giving away the wrong private key to external services (which would be the default if there is only one key pair, so nothing worse). 

Regarding the quantum resistance question, I'm not sure about the impact here.